torch_xla/experimental/custom_kernel.py

```diff
@@ -160,47 +160,184 @@ def wrapped_kernel(kernel: Callable,
+    # l and m that is needed for the backward. Then we lose all the shape checks.
+    # TODO: replicate the shape checks on flash_attention.
+    _flash_attention_impl = make_kernel_from_pallas(_flash_attention_impl,
+                                                    shape_dtype)
+    with torch.no_grad():
+      o = _flash_attention_impl(
+          q,
+          k,
+          v,
+          None,
+          None,
+          save_residuals,
+          causal,
+          1.0,
+          min(FlashAttention.DEFAULT_BLOCK_SIZES["block_b"], q.shape[0]),
+          min(FlashAttention.DEFAULT_BLOCK_SIZES["block_q"], q.shape[2]),
+          min(FlashAttention.DEFAULT_BLOCK_SIZES["block_k_major"], k.shape[2]),
+          min(FlashAttention.DEFAULT_BLOCK_SIZES["block_k"], k.shape[2]),
+          False,
+          static_argnums=range(5, 13))
+      if not save_residuals:
+        return o
+      o, *aux = o
+      l, m = (v[..., 0] for v in aux[-2:])
+
+    ctx.save_for_backward(q, k, v, o, l, m)
+    return o
+
+  @staticmethod
+  def backward(ctx, grad_output):
+    # Import JAX within the function such that we don't need to call the jax_import_guard()
+    # in the global scope which could cause problems for xmp.spawn.
+    jax_import_guard()
+    from jax.experimental.pallas.ops.tpu.flash_attention import _flash_attention_bwd_dq, _flash_attention_bwd_dkv
+
+    q, k, v, o, l, m = ctx.saved_tensors
+    causal = ctx.causal
+    grad_q = grad_k = grad_v = None
+
+    grad_i = torch.sum(
+        o.to(torch.float32) * grad_output.to(torch.float32),
+        axis=-1)  # [batch_size, num_heads, q_seq_len]
+
+    expanded_l = l.unsqueeze(-1).expand(3, 2, 128,
+                                        FlashAttention.MIN_BLOCK_SIZE)
+    expanded_m = m.unsqueeze(-1).expand(3, 2, 128,
+                                        FlashAttention.MIN_BLOCK_SIZE)
+    expanded_grad_i = grad_i.unsqueeze(-1).expand(3, 2, 128,
+                                                  FlashAttention.MIN_BLOCK_SIZE)
+
+    if ctx.needs_input_grad[0]:
+      payload, _ = trace_pallas(
+          _flash_attention_bwd_dq,
+          q,
+          k,
+          v,
+          None,
+          None,
+          l,
+          m,
+          grad_output,
+          grad_i,
+          block_q_major=min(FlashAttention.DEFAULT_BLOCK_SIZES["block_q_dq"],
+                            q.shape[2]),
+          block_k_major=min(
+              FlashAttention.DEFAULT_BLOCK_SIZES["block_k_major_dq"],
+              k.shape[2]),
+          block_k=min(FlashAttention.DEFAULT_BLOCK_SIZES["block_k_dq"],
+                      k.shape[2]),
+          sm_scale=1.0,
+          causal=causal,
+          mask_value=FlashAttention.DEFAULT_MASK_VALUE,
+          debug=False,
+          static_argnames=[
+              "block_q_major", "block_k_major", "block_k", "sm_scale", "causal",
+              "mask_value", "debug"
+          ])
+      grad_q = torch.empty(q.shape, dtype=q.dtype).to(xm.xla_device())
```

Comment:
it should be fine for the most part, but I think it is better to do `.to(q.deivce)` instead of move to `xm.xla_device()`. Should we add a check somewhere to make sure all tensors on XLA device?

Reply:
Fixed the .to(). For the second question, I guess Mosaic already guard it?